### DISCUSSION TITLE
Is variable shadowing considered good practice?

### DISCUSSION THREAD

**Discussion by AZMCode** [koka-lang/koka#481] [General]
I've been working on some code where using a mutable variable leads to it potentially escaping the function due to effect polymorphism, and thus I've resorted to shadowing the variable declarations instead.

This, however, leads to a compiler warning of shadowed variables, which led me to the question in the title.
Should I attempt to make the variable mutable, or just ignore the compiler warnings?

For context, this is the code I'm talking about, and the shadowed variable here is `out`:

```
pub fun (*%)<a,e>(
  a: a, b: a, ?modulo: a,
  ?(%): (a, a) -> e a,
  ?(+%): (a, a) -> e a,
  ?zero: a,
  ?bits: (a) -> e list<bool>
): e a
  val mod-a = a % ?modulo
  val mod-b = b % ?modulo
  val mod-b-bits = bits(mod-b)
  val last-index = length(mod-b-bits) - 1
  foldl(mod-b-bits.enumerate(), ?zero) fn(out,(i,bit))
    val out = 
      if bit then
        out +% mod-a
      else out
    val out =
      if i != last-index - 1 then
        out +% out
      else out
    out
```

**Comment by TimWhiting** [comment]
A local mutable variable wouldn't be bad for something like this, and I would expect it to compile close or basically to what you would expect in a C language. You could factor out the closure into a top level function which tends to work better than closures when dealing with problems with local variables and effect inference.

I think it is good practice to avoid the shadowing. Good names can make it clear when something is an intermediate result, an accumulator, or actually a finished product. As such here is one way you could refactor the code.

```koka
pub fun (*%)<a,e>(
  a: a, b: a, ?modulo: a,
  ?(%): (a, a) -> e a,
  ?(+%): (a, a) -> e a,
  ?zero: a,
  ?bits: (a) -> e list<bool>
): e a
  val mod-a = a % ?modulo
  val mod-b = b % ?modulo
  val mod-b-bits = bits(mod-b)
  val last-index = length(mod-b-bits) - 1
  foldl(mod-b-bits.enumerate(), ?zero) fn(acc,(i,bit))
    val out-bit = 
      if bit then acc +% mod-a 
      else acc
    if (i == last-index - 1) then
      out-bit +% out-bit
    else out-bit
```

Or you could do something a bit more drastic and factor out the transforming logic into your own custom `foldlt` which only does the transform on everything except the last element of the list. In fact I think this is a really clean solution especially since you don't have to create/allocaate an `enumerate`d list when all you really care about is whether you have hit the last element. Instead you take advantage of pattern matching ordering.
```koka
fun foldlt(list<a> ls, acc: b, t: (b) -> b, f: (b, a) -> e b): e b
  match ls
   Nil -> acc
   Cons(a, Nil) -> f(acc, a)
   Cons(a, ls') -> foldlt(ls', t(f(acc, a)), t, f)

pub fun (*%)<a,e>(
  a: a, b: a, ?modulo: a,
  ?(%): (a, a) -> e a,
  ?(+%): (a, a) -> e a,
  ?zero: a,
  ?bits: (a) -> e list<bool>
): e a
  val mod-a = a % ?modulo
  val mod-b = b % ?modulo
  val mod-b-bits = bits(mod-b)
  foldlt(mod-b-bits, ?zero, fn(out) out +% out) fn(acc,bit)
    if bit then acc +% mod-a 
    else acc
```
